### PR TITLE
WIP Add OpenGL support to libsimpleservo ...

### PR DESCRIPTION
**Pull request by Hyperion101010**
I have posted this first commit, just to take review that am i going on right track  ?

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #___ (GitHub issue number if applicable)

- [x] There are tests for these changes OR
- [x] These changes do not require tests because ___

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/22805)

### CODE REVIEW

**Comment by jdm** [comment]
@Hyperion101010 That looks like the right idea to me.

**Comment by paulrouget** [comment]
If you need help compiling/testing on Windows, Linux and/or Mac, let me know.
If you don't want our CI to build your PR, prefix the title of your PR with "WIP ".

We should probably add a simple example (under libsimplerservo/example/) to check if this will work correctly. I can look at this if it helps.

**Comment by Hyperion101010** [comment]
I guess there wasn't a test case checking the new changes,but the file compiled successfully though 
Any rough idea what we are going to do next ?

**Comment by Hyperion101010** [comment]
If it is solved , i am turning to https://github.com/servo/servo/issues/22760.
It was the soul purpose i tried fixing this first :)

**Comment by paulrouget** [comment]
The linux part doesn't look right to me.

You need to use glx (at least for now). See https://github.com/tomaka/glutin/blob/c16f997e4c6898bc3a3603c74c341c2a5497839c/src/platform/linux/x11.rs#L113

This is basically the hard part :)

The code I pointed to implement a `get_proc_address` method down the line (I think it's [here](https://github.com/tomaka/glutin/blob/c16f997e4c6898bc3a3603c74c341c2a5497839c/src/api/glx/mod.rs#L131]). You need to use that method inside `load_with`.

Basically, in the load_with method, you need to create a ffi::glx::Glx instance (as for EGL, you will need to create bindings. See how it's done [here](https://github.com/tomaka/glutin/blob/1ad6560860d37c1945cc9ffece645f66518f1d0f/src/api/glx/mod.rs#L24)) and call glx.GetProcAddress.

**Comment by paulrouget** [comment]
> If it is solved , i am turning to #22760.
> It was the soul purpose i wanted to fix this :)

That's great! I guess you could patch up something to see if your PR works with a libsimpleservo-based glutin port (it should work on Mac at least).

**Comment by Hyperion101010** [comment]
lets add for linux first , then i can atleast test myself on mu linux platform

**Comment by Hyperion101010** [comment]
hey paul you can now check the patch for linux as well , let me know that have i properly implemented GetProcAddressfunction ".

**Comment by paulrouget** [comment]
The linux code is not doing the right think. You need to dlopen the libgl. See https://github.com/tomaka/glutin/blob/c16f997e4c6898bc3a3603c74c341c2a5497839c/src/platform/linux/x11.rs#L42 - And use this instance of glx to call GetProcAddress.

File: ports/libsimpleservo/api/src/gl_glue.rs [hunk 1]
```diff
@@ -77,10 +81,122 @@ pub mod egl {
...
+
+        let mut opengl_dlopen =
+            unsafe { dlopen::dlopen(b"libGL.so.1\0".as_ptr() as *const _, dlopen::RTLD_NOW) };
+        if opengl_dlopen.is_null() {
+            unsafe {
+                opengl_dlopen = dlopen::dlopen(b"libGL.so\0".as_ptr() as *const _, dlopen::RTLD_NOW)
+            };
+        }
+
+        let opengl_pointer = ffi::glx::Glx::load_with(|addr| {
```
Comment (paulrouget) [inline ports/libsimpleservo/api/src/gl_glue.rs]: I'm surprised this compiles. Does `Glx::load_with` return a `Gl` object?

File: ports/libsimpleservo/api/src/gl_glue.rs [hunk 2]
```diff
@@ -77,10 +81,122 @@ pub mod egl {
...
+            let addr_name = CString::new(addr).unwrap();
+            unsafe { ffi::glx::Glx::GetProcAddress(addr_name.as_ptr() as *const _) as *const _ }
```
Comment (paulrouget) [inline ports/libsimpleservo/api/src/gl_glue.rs]: You can't just use `Glx` like that. You need to initialize it from the opened library.

See https://github.com/tomaka/glutin/blob/c16f997e4c6898bc3a3603c74c341c2a5497839c/src/platform/linux/x11.rs#L54

I haven't tested, but it should look like this:

```
let glx = Glx::load_with(|sym| {
  let sym = CString::new(sym).unwrap();
  unsafe { dlopen::dlsym(libglx, sym.as_ptr()) }
});

opengl_pointer = gl::GlFns::load_with(| addr|  glx.GetProcAddress(addr))
```

**Review by paulrouget — CHANGES_REQUESTED** [review]
I have the feeling you're a bit unsure about the general mechanic here.

Let me summarize a bit how OpenGL is loaded:

`GLFns::load_with` returns a GL object.
The GL object is initialized from a callback (the one passed to `load_with`). Basically, This callback is called with a name of a GL function, and the callback needs to return a pointer to that GL function. So in that callback, we need to find, in memory, the address of the symbol.

So there are 2 things to do: first make sure the GL library is loaded (which will make all the symbol available in memory), second, find the address of the symbol.

To load, we use `dlopen("mylib.so")`. To find the address, we usually use `dlsym`. But in the case of GL, there's a helper function, called `GetProcAddress`. But loading a library doesn't magically make GetProcAddress available. `dlsym` is needed or, in the case of GLX, there's a helper: `Glx::load_with`.

Hope this clarifies a bit :)

**Comment by paulrouget** [comment]
> About extra bindings paul i haven't researched it, but i read that it might be required for some other specialized frameworks that req OpenGl.
> But anyways it is fair that we neglect it

The problem is that you declare it but never use it.

**Comment by Hyperion101010** [comment]
I am closing this due to lack of my knowledge , I'll fix some simple bugs work on rust for a while .
In future I might try it . And sorry for any inconvenience caused 
See ya
